MacOS follow-up after 262487@main

pvollan opened the pull request:
#### fc633e7073a21283e1b912abb788ca0f60c4b5b9
<pre>
MacOS follow-up after 262487@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=254922">https://bugs.webkit.org/show_bug.cgi?id=254922</a>
rdar://107560995

Reviewed by Eric Carlson.

MacOS should also forward the caption bitmaps to the UI process, as was done for iOS in 262487@main.
This patch adds the required pieces to a new VideoFullscreenCaptions class, shared by iOS and macOS.
This patch also fixes a null pointer crash when entering fullscreen on macOS. Additionally, the
setting BlockMediaLayerRehostingInWebContentProcess is changed to only take effect when UI side
compositing is enabled, since this is the case it is intended for.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/VideoFullscreenCaptions.h:
* Source/WebCore/platform/cocoa/VideoFullscreenCaptions.mm:
(WebCore::VideoFullscreenCaptions::setTrackRepresentationImage):
(WebCore::VideoFullscreenCaptions::setTrackRepresentationContentsScale):
(WebCore::VideoFullscreenCaptions::setTrackRepresentationHidden):
(WebCore::VideoFullscreenCaptions::captionsLayer):
(WebCore::VideoFullscreenCaptions::setCaptionsFrame):
(WebCore::VideoFullscreenCaptions::setupCaptionsLayer):
(WebCore::VideoFullscreenCaptions::removeCaptionsLayer):
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(VideoFullscreenInterfaceAVKit::textTrackRepresentationUpdate): Deleted.
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetContentsScale): Deleted.
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetHidden): Deleted.
(VideoFullscreenInterfaceAVKit::captionsLayer): Deleted.
(VideoFullscreenInterfaceAVKit::setCaptionsFrame): Deleted.
(VideoFullscreenInterfaceAVKit::setupCaptionsLayer): Deleted.
(VideoFullscreenInterfaceAVKit::removeCaptionsLayer): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/platform/mac/VideoFullscreenInterfaceMac.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createLayerHostViewWithID):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationUpdate):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetContentsScale):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetHidden):
(WebKit::VideoFullscreenManagerProxy::didCleanupFullscreen):
(WebKit::VideoFullscreenManagerProxy::setVideoLayerFrame):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
(WebKit::VideoFullscreenInterfaceContext::setRootLayer): Deleted.
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenInterfaceContext::setRootLayer):
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::didExitFullscreen):
(WebKit::VideoFullscreenManager::didCleanupFullscreen):
(WebKit::VideoFullscreenManager::setupRemoteLayerHosting):

Canonical link: <a href="https://commits.webkit.org/262591@main">https://commits.webkit.org/262591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/493cd80f9e9d17acc72cc802193006de6c0e01b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1993 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2889 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2049 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2038 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1808 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1981 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1769 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2735 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1759 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1739 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1654 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1805 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1786 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2915 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1817 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1617 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1744 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1750 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/488 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1908 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/cocoa/VideoFullscreenCaptions.h:
```diff
@@ -0,0 +1,51 @@
+public:
+    WEBCORE_EXPORT void textTrackRepresentationUpdate(PlatformImagePtr textTrack);
+    WEBCORE_EXPORT void textTrackRepresentationSetContentsScale(float scale);
+    WEBCORE_EXPORT void textTrackRepresentationSetHidden(bool hidden);
```
Nit: I think the naming pattern should be verb first like the other half of the class, so `updateTrackRepresentation` (or `updateTrackRepresentationImage`?), setTrackRepresentationContentsScale`, and `setTrackRepresentationHidden`

pvollan replied:
> Nit: I think the naming pattern should be verb first like the other half of the class, so `updateTrackRepresentation` (or `updateTrackRepresentationImage`?), setTrackRepresentationContentsScale`, and `setTrackRepresentationHidden`

Fixed.

eric-carlson commented on Source/WebCore/platform/cocoa/VideoFullscreenCaptions.mm:
```diff
@@ -0,0 +1,81 @@
+CALayer *VideoFullscreenCaptions::captionsLayer()
+{
+    if (!m_captionsLayer)
+        m_captionsLayer = adoptNS([[CALayer alloc] init]);
```
Nit: it might be helpful for later debugging to set the layer's name

pvollan replied:
Added layer name.

eric-carlson commented on Source/WebCore/platform/cocoa/VideoFullscreenCaptions.mm:
```diff
@@ -0,0 +1,81 @@
+    [captionsLayer() removeFromSuperlayer];
+    [parent addSublayer:captionsLayer()];
+    
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
```
Is there a reason to not disable actions before removing the layer from its parent?

pvollan replied:
That is a good point. I included that in the transaction in the latest version.

Thanks for reviewing!

pvollan committed:
MacOS follow-up after 262487@main
https://bugs.webkit.org/show_bug.cgi?id=254922
rdar://107560995

Reviewed by Eric Carlson.

MacOS should also forward the caption bitmaps to the UI process, as was done for iOS in 262487@main.
This patch adds the required pieces to a new VideoFullscreenCaptions class, shared by iOS and macOS.
This patch also fixes a null pointer crash when entering fullscreen on macOS. Additionally, the
setting BlockMediaLayerRehostingInWebContentProcess is changed to only take effect when UI side
compositing is enabled, since this is the case it is intended for.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/VideoFullscreenCaptions.h:
* Source/WebCore/platform/cocoa/VideoFullscreenCaptions.mm:
(WebCore::VideoFullscreenCaptions::setTrackRepresentationImage):
(WebCore::VideoFullscreenCaptions::setTrackRepresentationContentsScale):
(WebCore::VideoFullscreenCaptions::setTrackRepresentationHidden):
(WebCore::VideoFullscreenCaptions::captionsLayer):
(WebCore::VideoFullscreenCaptions::setCaptionsFrame):
(WebCore::VideoFullscreenCaptions::setupCaptionsLayer):
(WebCore::VideoFullscreenCaptions::removeCaptionsLayer):
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(VideoFullscreenInterfaceAVKit::textTrackRepresentationUpdate): Deleted.
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetContentsScale): Deleted.
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetHidden): Deleted.
(VideoFullscreenInterfaceAVKit::captionsLayer): Deleted.
(VideoFullscreenInterfaceAVKit::setCaptionsFrame): Deleted.
(VideoFullscreenInterfaceAVKit::setupCaptionsLayer): Deleted.
(VideoFullscreenInterfaceAVKit::removeCaptionsLayer): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/platform/mac/VideoFullscreenInterfaceMac.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createLayerHostViewWithID):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationUpdate):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetContentsScale):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetHidden):
(WebKit::VideoFullscreenManagerProxy::didCleanupFullscreen):
(WebKit::VideoFullscreenManagerProxy::setVideoLayerFrame):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
(WebKit::VideoFullscreenInterfaceContext::setRootLayer): Deleted.
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenInterfaceContext::setRootLayer):
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::didExitFullscreen):
(WebKit::VideoFullscreenManager::didCleanupFullscreen):
(WebKit::VideoFullscreenManager::setupRemoteLayerHosting):

Canonical link: https://commits.webkit.org/262591@main
